django/core/management/commands/dumpdata.py

```diff
@@ -6,6 +8,40 @@
 from django.core.management.utils import parse_apps_and_model_labels
 from django.db import DEFAULT_DB_ALIAS, router
 
+try:
+    import bz2
+    has_bz2 = True
+except ImportError:
+    has_bz2 = False
+
+try:
+    import lzma
+    has_lzma = True
+except ImportError:
+    has_lzma = False
+
+
+def get_open_method(file_path):
+    """Return open method based on the file extension."""
+    file_root, file_ext = os.path.splitext(file_path)
+    compression_formats = {
+        '.gz': (gzip.open, {}, file_path),
+        '.bz2': (open, {}, file_root),
+        '.lzma': (open, {}, file_root),
+        '.xz': (open, {}, file_root),
+    }
+    if has_bz2:
+        compression_formats['.bz2'] = (bz2.open, {}, file_path)
+    if has_lzma:
+        compression_formats['.lzma'] = (
+            lzma.open, {'format': lzma.FORMAT_ALONE}, file_path
+        )
+        compression_formats['.xz'] = (lzma.open, {}, file_path)
+    try:
+        return compression_formats[file_ext]
+    except KeyError:
+        return (open, {}, file_path)
```

Comment:
Given the discussion about removing the extension when e.g. `not has_bz2`, should this return `file_root` rather than `file_path`? (The adjustment and warning are untested yes?)

Reply:
By default I've set `file_root` in `compression_formats` for `gz`, `lzma` and `xz` to remove the extension if your Python interpreter doesn't support those formats.

This line instead continues to work as it's now and writes plain text when you specify an extension that is not explicitly supported (e.g. `myfixtures.myextension`).

I'm working on testing this case but I'm trying to figure out how to emulate a Python interpreter with lacking support for `bzip2` or `lzma`.

Reply:
I've added a test for the '.zip' extension removing and related warning.

Reply:
OK, thanks @pauloxnet

> I'm working on testing this case but I'm trying to figure out how to emulate a Python interpreter with lacking support for bzip2 or lzma.

Update the module level variables in setUp, and restore in tearDown?

Reply:
Actually I tested the filename adjustment and related warning for zip extension so I don't think it' necessary anymore forcing module level variables.